src/CodeXMLParse.cpp

```diff
@@ -162,11 +162,30 @@ void AnnotateConstantVariable(Varnode *varnode, std::vector<RCodeAnnotation> *ou
 	out->push_back(annotation);
 }
 
+// Annotates local variables and function parameters
+void AnnotateLocalVariable(pugi::xml_node node, std::vector<RCodeAnnotation> *out)
+{
+	pugi::xml_attribute attr = node.attribute("color");
+	if (attr.empty())
+		return;
+	std::string color = attr.as_string();
+	if (color == "")
+		return;
+	
+	RCodeAnnotation annotation = {};
+	annotation.variable.name = strdup(node.child_value());
+	if (color == "param")
+		annotation.type = R_CODE_ANNOTATION_TYPE_FUNCTION_PARAMETER;
+    else if (color == "var")
+		annotation.type = R_CODE_ANNOTATION_TYPE_LOCAL_VARIABLE;
+	out->push_back(annotation);
```

Comment:
Please do not make decisions based on color.

Reply:
@karliss @thestr4ng3r  I had seen the decision that leads to color. For PR #124 , I have used info from here. https://github.com/thestr4ng3r/ghidra/blob/6c10f36f06468f866188cccf960c019779fb9028/Ghidra/Features/Decompiler/src/decompile/cpp/variable.cc#L452-L470

In there also, the same logic (if I didn't miss out anything) is used there to classify local variables and parameters. This is the reason why I used color to classify a variable as a parameter and local variable. Have you seen examples where color is mistakenly annotated?

Reply:
Also note that in some cases, the XML has just the color of the variable e.g. ``` <variable color="param">param_1</variable> ```
In other cases where variable has an attribute `varref`, it is first checked to make sure that it's not a global variable or a constant variable.

Reply:
Earlier, forgot to add the following check, I have added it now.
```
	else (!varnode->getHigh()->isPersist() && (varnode->getHigh()->getSymbol() != (Symbol *)0))
		AnnotateLocalVariable(node, out);
```

Reply:
Do you know the place in `printc.cc` where these cases without `varref` like `<variable color="param">param_1</variable>` are emitted?

Reply:
@thestr4ng3r I'm not sure if I understand the question. If you were asking to confirm if `<variable color="param">param_1</variable>` is given the color="param" from the location you pointed to, then yes it happens at the place [you linked.](https://github.com/thestr4ng3r/ghidra/blob/master/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc#L1560-L1563)

Reply:
Yes, that is what I meant. Now the question is why exactly some variables don't get the `varref` attribute.

Reply:
@thestr4ng3r I had noticed that it doesn't get the `varref` attribute when in places where variables are declared. For reference, let's use [example_pdgx_pdg_output.zip](https://github.com/radareorg/r2ghidra-dec/files/5006086/example_pdgx_pdg_output.zip). In the case of local variable `pvVar2`, it's used in three places.
Declaration.
`<type color="type" id="0x80feab8523c497cf">void</type>
        <syntax></syntax>
        <op>*</op>
        <syntax></syntax>
        <variable color="var">pvVar2</variable>`
Rest
1. `<variable color="var" varref="0x3c4" opref="0x27d">pvVar2</variable>`
2. `<variable color="var" varref="0x3c4" opref="0x1d">pvVar2</variable>`

Reply:
@thestr4ng3r Do you agree that the current code will work in cases where we have `varref`? From what I understand, in cases where we have varref, we are doing exactly what's done in [variable.cc(#L452-L470)](https://github.com/thestr4ng3r/ghidra/blob/6c10f36f06468f866188cccf960c019779fb9028/Ghidra/Features/Decompiler/src/decompile/cpp/variable.cc#L452-L470
) when we have the check
```
	else if (!varnode->getHigh()->isPersist() && (varnode->getHigh()->getSymbol() != (Symbol *)0))
		AnnotateLocalVariable(node, out);
```

Reply:
Ah I see. Yes it would work but using color is still conceptually wrong. Moreover it currently uses the printed text `node.child_value()` as the variable name which is only marginally better than just "grepping".
What you do get however in the case of the declaration is the following:
```
<vardecl symref="0x400001000000000d">
        <type color="type" id="0x80feab8523c497cf">void</type>
        <syntax></syntax>
        <op>*</op>
        <syntax></syntax>
        <variable color="var">pvVar2</variable>
</vardecl>
```
So you can get the true info about the symbol itself from the parent's `symref`:
```
<symbol name="pvVar2" id="0x400001000000000d" cat="-1">
...
```
which also has the `cat` field indicating whether it is a parameter or regular variable.

Reply:
Thanks a lot, this is really helpful. Here's what I am planning to do now.
Make a Symbol_ID -> Symbol map.

1. If varref exists, get Symbol from the varnode by `varnode->getHigh()->getSymbol()`.
2. Otherwise, get symbol from the map.

In order to get the `symref`, I will do `node.parent()` which will give the `vardecl` node and then get the attribute `symref`.

Reply:
Yep, that approach sounds good!

Reply:
@thestr4ng3r @karliss I have pushed changes now.